clang/test/SemaCXX/warn-unsafe-buffer-usage-count-attributed-pointer-argument.cpp

```diff
@@ -665,6 +685,30 @@ static void previous_infinite_loop3(int * __counted_by(n + n * m) p, size_t n,
   previous_infinite_loop3(p, n, q, r, m, o + 1); // expected-warning 2{{unsafe assignment to function parameter of count-attributed type}}
 }
 
+// Check nullable variants.
+
+void nonnull_tofrom_nullable(size_t n,
+                             int *__counted_by(n) cb,
+                             int *__counted_by_or_null(n) cbn,
+                             void *__sized_by(n) sb,
+                             void *__sized_by_or_null(n) sbn) {
+  cb_int(cb, n);
+  // __counted_by_or_null(count) can have arbitrary count if the pointer is
+  // null. Since __counted_by() does not allow this, we diagnose
+  // __counted_by_or_null() -> __counted_by().
+  cb_int(cbn, n); // expected-warning{{unsafe assignment to function parameter of count-attributed type}}
+
+  cbn_int(cb, n);
+  // __counted_by() -> __counted_by_or_null() is fine.
+  cbn_int(cbn, n);
+
+  sb_void(sb, n);
+  sb_void(sbn, n); // expected-warning{{unsafe assignment to function parameter of count-attributed type}}
+
+  sbn_void(sb, n);
+  sbn_void(sbn, n);
+}
```

Comment:
Can we add more tests:
- unsafe cases of  `cbn_int` and `sun_void`;
- passing `__counted/sized_by_or_null` pointers to `std::span`
- passing `__counted/sized_by_or_null` pointers to the 1st and 2nd parameters of `snprintf` functions